gcp_variant_transforms/transforms/infer_headers.py

```diff
@@ -223,6 +248,71 @@ def _infer_info_fields(self, variant, defined_headers):
               str(defined_header.get(_HeaderKeyConstants.NUM)),
               str(info_field_value), str(variant))
           infos[info_field_key] = corrected_info
+
+  def _infer_type_info_fields(self, variant, infos, defined_headers):
+    # type: (vcfio.Variant, Dict[str, Info], vcf_header_io.VcfHeader) -> None
+    """Updates `infos` with inferred annotation type info fields.
+
+    All annotation headers in each annotation field are converted to Info header
+    lines where the new ID corresponds to the given annotation field and header,
+    and the new TYPE corresponds to inferred type of the original header. Since
+    each variant potentially contains multiple values for each annotation
+    header, a small 'merge' of value types is performed before VcfHeader
+    creation for each variant.
+    Args:
+      variant: variant obj.
+      infos: dict of (info_key(str), :class:`Info`) for any info field in
+        `variant` that is not defined in the header or the definition mismatches
+        the field values.
+      defined_headers: header fields defined in header section of VCF files.
+    """
+    resolver = vcf_field_conflict_resolver.FieldConflictResolver(
+        resolve_always=True)
+    for field in self._annotation_fields:
+      if field not in variant.info:
+        continue
+      annotation_headers = annotation_parser.extract_annotation_names(
+          defined_headers.infos[field][
+              vcf_header_io.VcfParserHeaderKeyConstants.DESC], with_alt=True)
+      annotation_values = [annotation_parser.extract_annotation_list_with_alt(
+          annotation) for annotation in variant.info[field]]
+      annotation_values = zip(*annotation_values)
+      for header, values in zip(annotation_headers, annotation_values):
+        variant_merged_type = vcf_header_io.VcfHeaderFieldTypeConstants.INTEGER
+        for v in values:
+          if v != '':
+            variant_merged_type = resolver.resolve_attribute_conflict(
+                vcf_header_io.VcfParserHeaderKeyConstants.TYPE,
+                variant_merged_type,
+                self._get_field_type(v))
+        key_id = vcf_header_io.BASE_TYPE_KEY.format(field, header)
+        infos[key_id] = Info(key_id,
+                             1,
+                             variant_merged_type,
+                             '',  # NO_DESCRIPTION
```

Comment:
I think we can have a description because these will be saved in the merged header that we generate, correct? E.g., something like 'The inferred type field for annotation X'.